src/CodeGen_Internal.cpp

```diff
@@ -388,6 +388,24 @@ Expr lower_int_uint_mod(const Expr &a, const Expr &b) {
     }
 }
 
+Expr ulong_div(Expr num, Expr den) {
+    internal_assert(num.type() == den.type());
```

Comment:
I think it would be a big optimization to enable the result is a smaller type than the inputs. For example, int32 / int16 -> int16 would be much faster to compute than int32 / int16 -> int32.

Reply:
Right. I'll add some bounds analysis check for ranges of num and den and perform operations in lowest type that num and den can be represented in. Also will add the bounds analysis check to see if some iterations can be skipped.